tests/src/main/java/net/neoforged/neoforge/debug/capabilities/handlers/resources/ComponentResourceTests.java

```diff
@@ -0,0 +1,255 @@
+import net.neoforged.testframework.gametest.EmptyTemplate;
+import net.neoforged.testframework.gametest.ExtendedGameTestHelper;
+import net.neoforged.testframework.gametest.GameTest;
+
+// Tests component storages on ItemStacks as well as the handler logic for those components. Also validates the Codec and Stream codec
+@ForEachTest(groups = ResourceHandlerTestSetup.GROUP_ID, idPrefix = "resource.handler.component.")
+public class ComponentResourceTests {
+    @GameTest
+    @EmptyTemplate
+    @TestHolder(description = "Tests that FluidHandlerItemStack works")
+    public static void testFluidHandlerItemStack(ExtendedGameTestHelper helper) {
+        Player player = helper.makeMockPlayer();
+        player.setItemInHand(InteractionHand.MAIN_HAND, Items.APPLE.getDefaultInstance());
+        IItemContext context = PlayerItemContext.ofHand(player, InteractionHand.MAIN_HAND);
+        int capacity = 2 * FluidType.BUCKET_VOLUME;
+
+        ItemContextFluidHandler.Consumable handler = new ItemContextFluidHandler.Consumable(context, ResourceHandlerTestSetup.Content.SINGLE_FLUID_CONTENT.get(), capacity);
+
+        if (handler.size() != 1)
+            helper.fail("Expected a single tank");
+
+        if (handler.getCapacity(0, FluidResource.EMPTY) != capacity)
+            helper.fail("Expected tank capacity of " + capacity);
+
+        if (handler.getAmount(0) != 0)
+            helper.fail("Expected empty tank");
+        try (Transaction tx = TransactionManager.open(null)) {
+            int inserted = handler.insert(0, FluidResource.of(Fluids.WATER), FluidType.BUCKET_VOLUME, tx);
+            if (inserted != FluidType.BUCKET_VOLUME)
+                helper.fail("Expected to be able to fill a bucket of water");
+            tx.commit();
+        }
+
+        if (!player.getMainHandItem().has(ResourceHandlerTestSetup.Content.SINGLE_FLUID_CONTENT))
+            helper.fail("Expected fluid stack component");
+
+        if (!ResourceHandlerUtil.resourceAndCountMatches(handler, 0, FluidResource.of(Fluids.WATER), FluidType.BUCKET_VOLUME))
+            helper.fail("Expected a bucket of water");
+
+        try (Transaction tx = TransactionManager.open(null)) {
+            int extracted = handler.extract(0, FluidResource.of(Fluids.WATER), FluidType.BUCKET_VOLUME, tx);
+            if (extracted != FluidType.BUCKET_VOLUME)
+                helper.fail("Expected to drain a bucket of water");
+            tx.commit();
+        }
+
+        if (!ResourceHandlerUtil.isIndexEmpty(handler, 0))
+            helper.fail("Expected empty tank");
+
+        ResourceStack<FluidResource> component = player.getMainHandItem().get(ResourceHandlerTestSetup.Content.SINGLE_FLUID_CONTENT);
+
+        if (component == null || !component.isEmpty())
+            helper.fail("Fluid stack component not found, or was empty");
+
+        helper.succeed();
+    }
+
+    @GameTest
+    @EmptyTemplate
+    @TestHolder(description = "Tests that FluidHandlerItemStack works")
```

Comment:
This should more accurately describe the test at hand; looking at the code it should include something like "inserting and draining fluids in a single transaction should succeed"